fsharp-backend/src/LibExecution/Interpreter.fs

```diff
@@ -48,6 +49,279 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
+                   let id = Pattern.toID pat
+                   state.tracing.traceDval false id (incomplete id))
+                 args
+             })
+
+      let! matchVal = eval state st matchExpr
+
+      do!
+        Ply.List.iterSequentially
+          (fun (pattern, expr) -> matchAndExecute matchVal [] (pattern, expr))
+          cases
+
+      return matchResult.Value
+    }
+
+  let evalMatchNew patternId matchExpr (cases : List<Pattern * Expr>) =
+    let rec patternPasses dv pattern : bool =
+      match pattern with
+      | PInteger (_, i) -> dv = DInt i
+      | PBool (_, b) -> dv = DBool b
+      | PCharacter (_, c) -> dv = DChar c
+      | PString (_, str) -> dv = DStr str
+      | PFloat (_, v) -> dv = DFloat v
+      | PNull (_) -> dv = DNull
+      | PVariable (_, _) -> not (Dval.isFake dv)
+      | PBlank (_) -> false
+      | PConstructor (_, name, args) ->
+        match (name, args, dv) with
+        | "Nothing", [], DOption None -> true
+
+        | "Just", [ p ], DOption (Some _v)
+        | "Ok", [ p ], DResult (Ok _v)
+        | "Error", [ p ], DResult (Error _v) -> patternPasses dv p
+
+        | "Nothing", [], _ -> false
+        | _ -> false
+
+    let tracePattern onExecutionPath dv pattern : List<string * Dval> =
+      let traceFn (id, dv) = state.tracing.traceDval onExecutionPath id dv
+      let traceIncomplete id = state.tracing.traceDval false id (incomplete id)
+
+      let rec r dv pattern =
+        match pattern with
+        | PInteger (id, i) ->
+          traceFn (id, DInt i)
+          []
+        | PBool (id, b) ->
+          traceFn (id, DBool b)
+          []
+        | PCharacter (id, c) ->
+          traceFn (id, DChar c)
+          []
+        | PString (id, s) ->
+          traceFn (id, DStr s)
+          []
+        | PFloat (id, f) ->
+          traceFn (id, DFloat f)
+          []
+        | PNull (id) ->
+          traceFn (id, DNull)
+          []
+        | PBlank (id) ->
+          traceFn (id, incomplete id)
+          []
+        | PVariable (id, varName) ->
+          traceFn (id, dv)
+          [ (varName, dv) ]
+        | PConstructor (id, name, args) ->
+          match (name, args, dv) with
+          | "Nothing", [], _ ->
+            traceFn (id, DOption None)
+            []
+
+          | "Just", [ p ], DOption (Some v)
+          | "Ok", [ p ], DResult (Ok v)
+          | "Error", [ p ], DResult (Error v) when patternPasses v p ->
+            traceFn (id, dv)
+
+            r v p
+
+          | _name, subPatterns, _ ->
+            traceIncomplete id
+
+            subPatterns
+            |> List.map Pattern.toID
+            |> List.iter (fun pId -> traceFn (pId, incomplete pId))
+
+            []
+
+      r dv pattern
+
+    uply {
+      // The value we're matching against
+      let! matchVal = eval state st matchExpr
+
+      let stWithNewVars newVars = Map.mergeFavoringRight st (Map.ofList newVars)
+
+      let mutable foundMatch : Option<Dval> = None
```

Comment:
Perhaps `hasMatched`? I was confused by `foundMAtch`.

Reply:
`hasMatched` feels like a `bool` to me, and I was trying to avoid having 2 mutable bindings (instead, just using one `option`).

I've renamed to `resultMaybe` - I think that's a clearer compromise?

Reply:
I was going the other way – I wanted to have a variable name where are the primary focus was on tracking whether the pattern had matched for the first time yet.

Reply:
`foundMatch` was aimed towards that goal - it's the found match, it being None is a pretty clear message that a match hasn't been found?